File: client/Packages/IMCore/Sources/IMCore/Services/ConversationService.swift
import Foundation

@MainActor
public class ConversationService {
    public static let shared = ConversationService()
    private let api = APIClient.shared

    private init() {}

    // MARK: - List
    public func listConversations(page: Int = 1, size: Int = 20) async throws -> [ConvListItem] {
        let result: PaginatedData<ConvListItem> = try await api.request(
            "/api/v1/conversations",
            query: ["page": "\(page)", "size": "\(size)"]
        )
        return result.items
    }

    // MARK: - Detail
    public func getConversationDetail(convID: String) async throws -> ConversationDetail {
        let detail: ConversationDetail = try await api.request("/api/v1/conversations/\(convID)")
        return detail
    }

    // MARK: - Create P2P
    public func createP2P(userID: String) async throws -> (convID: String, name: String) {
        struct P2PReq: Codable, Sendable {
            let userID: String
            enum CodingKeys: String, CodingKey {
                case userID = "user_id"
            }
        }
        struct P2PResp: Codable, Sendable {
            let convID: String
            let name: String
            let type: ConvType
            enum CodingKeys: String, CodingKey {
                case convID = "conv_id"
                case name, type
            }
        }
        let resp: P2PResp = try await api.request(
            "/api/v1/conversations/p2p",
            method: .post,
            body: P2PReq(userID: userID)
        )
        return (resp.convID, resp.name)
    }

    // MARK: - Create Group
    public func createGroup(name: String, memberIDs: [String]) async throws -> Conversation {
        let detail: Conversation = try await api.request(
            "/api/v1/conversations/group",
            method: .post,
            body: CreateGroupReq(name: name, memberIDs: memberIDs)
        )
        return detail
    }

    // MARK: - Mark Read
    public func markRead(convID: String, msgID: Int64) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/read",
            method: .post,
            body: MarkReadReq(msgID: msgID)
        )
    }

    // MARK: - Members
    public func addMembers(convID: String, userIDs: [String]) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/members",
            method: .post,
            body: AddMembersReq(userIDs: userIDs)
        )
    }

    public func removeMember(convID: String, userID: String) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/members/\(userID)",
            method: .delete
        )
    }

    public func leaveGroup(convID: String) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/leave",
            method: .post
        )
    }

    // MARK: - Group Search
    public func searchGroups(query: String, page: Int = 1, size: Int = 20) async throws -> [GroupSearchItem] {
        let result: PaginatedData<GroupSearchItem> = try await api.request(
            "/api/v1/groups/search",
            query: ["q": query, "page": "\(page)", "size": "\(size)"]
        )
        return result.items
    }

    // MARK: - Join Requests
    public func requestJoin(convID: String) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/join-requests",
            method: .post
        )
    }

    public func listJoinRequests(convID: String) async throws -> [JoinRequest] {
        let requests: [JoinRequest] = try await api.request(
            "/api/v1/conversations/\(convID)/join-requests"
        )
        return requests
    }

    public func approveJoinRequest(convID: String, userID: String) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/join-requests/\(userID)/approve",
            method: .post
        )
    }

    public func rejectJoinRequest(convID: String, userID: String) async throws {
        let _: [String: String] = try await api.request(
            "/api/v1/conversations/\(convID)/join-requests/\(userID)/reject",
            method: .post
        )
    }

    // MARK: - Requests
    private struct CreateGroupReq: Codable, Sendable {
        let name: String
        let memberIDs: [String]

        enum CodingKeys: String, CodingKey {
            case name
            case memberIDs = "member_ids"
        }
    }

    private struct MarkReadReq: Codable, Sendable {
        let msgID: Int64

        enum CodingKeys: String, CodingKey {
            case msgID = "msg_id"
        }
    }

    private struct AddMembersReq: Codable, Sendable {
        let userIDs: [String]

        enum CodingKeys: String, CodingKey {
            case userIDs = "user_ids"
        }
    }
}

File: server/web/src/services/conversation-service.ts
import { api } from './api-client'
import type { ConvListItem, ConversationDetail, JoinRequest } from '@/types/conversation'
import type { PaginatedData } from '@/types/api'

export const conversationService = {
  list(page = 1, size = 50) {
    return api.request<PaginatedData<ConvListItem>>('/api/v1/conversations', {
      query: { page, size },
    })
  },

  getDetail(convId: string) {
    return api.request<ConversationDetail>(`/api/v1/conversations/${convId}`)
  },

  createP2P(userId: string) {
    return api.request<{ conv_id: string }>('/api/v1/conversations/p2p', {
      method: 'POST',
      body: { user_id: userId },
    })
  },

  createGroup(name: string, headline: string, memberIds: string[], avatar?: string) {
    return api.request<{ conv_id: string }>('/api/v1/conversations/group', {
      method: 'POST',
      body: { name, headline, member_ids: memberIds, avatar },
    })
  },

  updateGroup(convId: string, data: { name?: string; avatar?: string; notice?: string; cover?: string; headline?: string }) {
    return api.request<null>(`/api/v1/conversations/${convId}`, { method: 'PUT', body: data })
  },

  pin(convId: string) {
    return api.request<{ conv_id: string; pinned: boolean }>(`/api/v1/conversations/${convId}/pin`, { method: 'POST' })
  },

  unpin(convId: string) {
    return api.request<{ conv_id: string; pinned: boolean }>(`/api/v1/conversations/${convId}/unpin`, { method: 'POST' })
  },

  clone(convId: string, name?: string) {
    return api.request<{ conv_id: string; name: string }>(`/api/v1/conversations/${convId}/clone`, { method: 'POST', body: name ? { name } : {} })
  },

  addMembers(convId: string, userIds: string[]) {
    return api.request<null>(`/api/v1/conversations/${convId}/members`, {
      method: 'POST',
      body: { user_ids: userIds },
    })
  },

  removeMember(convId: string, userId: string) {
    return api.request<null>(`/api/v1/conversations/${convId}/members/${userId}`, {
      method: 'DELETE',
    })
  },

  leave(convId: string) {
    return api.request<null>(`/api/v1/conversations/${convId}/leave`, {
      method: 'POST',
    })
  },

  disband(convId: string) {
    return api.request<{ status: string }>(`/api/v1/conversations/${convId}/disband`, {
      method: 'POST',
    })
  },

  markRead(convId: string, msgId: number) {
    return api.request<null>(`/api/v1/conversations/${convId}/read`, {
      method: 'POST',
      body: { msg_id: msgId },
    })
  },

  requestJoin(convId: string) {
    return api.request<null>(`/api/v1/conversations/${convId}/join-requests`, {
      method: 'POST',
    })
  },

  listJoinRequests(convId: string) {
    return api.request<JoinRequest[]>(`/api/v1/conversations/${convId}/join-requests`)
  },

  approveJoinRequest(convId: string, userId: string) {
    return api.request<null>(`/api/v1/conversations/${convId}/join-requests/${userId}/approve`, {
      method: 'POST',
    })
  },

  rejectJoinRequest(convId: string, userId: string) {
    return api.request<null>(`/api/v1/conversations/${convId}/join-requests/${userId}/reject`, {
      method: 'POST',
    })
  },

  searchGroups(query: string) {
    return api.request<PaginatedData<ConvListItem>>('/api/v1/groups/search', { query: { q: query } }).then(d => d.items)
  },

  getUnreadTotal() {
    return api.request<{ total: number }>('/api/v1/conversations/unread/total')
  },

  getSettings(convId: string) {
    return api.request<{ settings: Record<string, unknown> }>(`/api/v1/conversations/${convId}/settings`)
  },

  updateSettings(convId: string, settings: Record<string, unknown>) {
    return api.request<{ settings: Record<string, unknown> }>(`/api/v1/conversations/${convId}/settings`, {
      method: 'PUT',
      body: { settings },
    })
  },
}
